torch_xla/csrc/convolution.cpp

```diff
@@ -9,232 +11,80 @@
+  // Stride, dilation and padding must be set for the batch and feature in the
+  // TF convolution metadata. Set them to 1 (stride and dilation) or 0 (padding)
+  // for the batch and feature dimensions.
+  conv_op_attrs.dilations = {1, 1};
+  std::copy(spatial_dilation.begin(), spatial_dilation.end(),
+            std::back_inserter(conv_op_attrs.dilations));
+  conv_op_attrs.strides = {1, 1};
+  std::copy(spatial_stride.begin(), spatial_stride.end(),
+            std::back_inserter(conv_op_attrs.strides));
+  conv_op_attrs.padding = tensorflow::Padding::EXPLICIT;
+  conv_op_attrs.explicit_paddings.resize(4);
+  for (int spatial_dim = 0; spatial_dim < num_spatial_dims; ++spatial_dim) {
+    conv_op_attrs.explicit_paddings.push_back(spatial_padding[spatial_dim]);
+    conv_op_attrs.explicit_paddings.push_back(spatial_padding[spatial_dim]);
   }
+  conv_op_attrs.data_format = tensorflow::TensorFormat::FORMAT_NCHW;
+  return conv_op_attrs;
+}
 
-  // Mirror the filter in the spatial dimensions.
-  xla::XlaOp mirrored_weights = xla::Rev(filter, kernel_spatial_dims);
+std::vector<xla::int64> FilterTransposePermutation() { return {2, 3, 1, 0}; }
 
-  // We'll need to undo the initial input padding once on the input backprop
-  // result since edges are constant and have to be discarded for the gradient.
-  xla::PaddingConfig padding_config;
-  for (int i = 0; i < 2; ++i) {
-    padding_config.add_dimensions();
-  }
-  for (int i = 0; i < 2; ++i) {
-    xla::PaddingConfig::PaddingConfigDimension* dims =
-        padding_config.add_dimensions();
-    dims->set_edge_padding_low(-padding_attr[i]);
-    dims->set_edge_padding_high(-padding_attr[i]);
-  }
-
-  // activation gradients
-  //   = gradients (with padding and dilation) <conv> mirrored_weights
+// Computes the input gradient for a convolution.
+xla::XlaOp BuildThnnConv2dBackwardInput(
+    const xla::XlaOp& grad_output, const xla::XlaOp& kernel,
+    const xla::Shape& input_shape,
+    tensorflow::gtl::ArraySlice<const xla::int64> spatial_stride,
+    tensorflow::gtl::ArraySlice<const xla::int64> spatial_padding) {
+  tensorflow::ConvOpAttrs conv_op_attrs =
+      MakeConvOpAttrs(spatial_stride, spatial_padding, {1, 1});
+  xla::XlaOp kernel_transposed =
+      xla::Transpose(kernel, FilterTransposePermutation());
   xla::PrecisionConfig precision_config =
       XlaHelpers::BuildPrecisionConfig(XlaHelpers::mat_mul_precision());
-  xla::Shape weight_shape = XlaHelpers::ShapeOfXlaOp(weight);
-  return xla::Pad(
-      xla::ConvGeneralDilated(grad_output, mirrored_weights,
-                              /*window_strides=*/ones, padding, lhs_dilation,
-                              rhs_dilation, dnums,
-                              /*feature_group_count=*/1,
-                              /*batch_group_count=*/1, &precision_config),
-      XlaHelpers::ScalarValue<float>(0, weight_shape.element_type(), builder),
-      padding_config);
+  return tensorflow::MakeXlaBackpropInputConvOp(
```

Comment:
Use our Consume(), which throws instead of dying.

Reply:
Done.